Zend/zend_enum.c

```diff
@@ -207,31 +207,21 @@ static ZEND_NAMED_FUNCTION(zend_enum_cases_func)
 	} ZEND_HASH_FOREACH_END();
 }
 
-static void zend_enum_from_base(INTERNAL_FUNCTION_PARAMETERS, bool try)
+ZEND_API zend_result zend_enum_get_case_by_value(zend_object **result, zend_class_entry *ce, zend_long long_key, zend_string *string_key, bool try)
```

Comment:
@iluuu1994: I just came across [replies in room 11](https://chat.stackoverflow.com/transcript/message/54523075#54523075) from earlier this week. My original question in the chat had more to do with why `zend_call_method` was not permitting me to invoke `BackedEnum::from()` and `tryFrom()` directly. It looks like an issue with the `static` return type validation, which I tried to dissect in [this comment](https://github.com/mongodb/mongo-php-driver/pull/1317/files#r866371513).

That said, if we're going to expose a new C API I think it is more generally useful to have a function that constructs an enum from its case, since that can be used for both pure _and_ backed enums. This is the approach I implemented in my `ext-mongodb` PR [here](https://github.com/mongodb/mongo-php-driver/pull/1317/files#diff-09bc6eb8abd9dcb66a40cd127006f752677a35440667f47d505cd9474abaff43R966). The code is derived from existing logic in `php_var_unserialize_internal`.

For context, in the MongoDB PHP driver, we already had a facility to allow PHP objects to be serialized to [BSON](https://bsonspec.org/) by recording the class name in an extra document field (i.e. `__pclass`). A recent bug report pointed out that this logic didn't work for enums. A pure and backed enum would encode to an object correctly (the class has a read-only `name` property, and backed enums also get a `value` property), but there was an issue on unserializing BSON back to a PHP object since they must be reconstructed specially. Since enums are also serializable in PHP, the approach in `php_var_unserialize_internal` proved useful.

My questions about being unable to invoke `from()` and `tryFrom()` were mostly academic, as the error just didn't make sense to me. But were I to use that solution, we'd be limited to only serializing BackedEnum instances. Therefore, the `php_var_unserialize_internal` reconstruction model is preferable.

Reply:
I see, these are two distinct issues.

1. Fetching cases by backed value requires a function call (why it didn't work for your case I can't say, didn't look at it closely). That's what this PR improves.
2. Fetching any constant (not just enums) by name internally requires quite a few hoops to jump through. That can probably be improve in general, and then extended for enums (to add asserts that the constant is actually an enum case).

IMO both would make sense.

Reply:
@jmikola I added improvement of the C class constant API to my list.